rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <sstream>
+
+#include "rmw/qos_profiles.h"
+
+#include "rmw_dds_common/qos.hpp"
+
+namespace rmw_dds_common
+{
+
+static bool
+operator==(rmw_time_t t1, rmw_time_t t2)
+{
+  return t1.sec == t2.sec && t1.nsec == t2.nsec;
+}
+
+static bool
+operator!=(rmw_time_t t1, rmw_time_t t2)
+{
+  return !(t1 == t2);
+}
+
+static bool
+operator<(rmw_time_t t1, rmw_time_t t2)
+{
+  if (t1.sec < t2.sec) {
+    return true;
+  } else if (t1.sec == t2.sec && t1.nsec < t2.nsec) {
+    return true;
+  }
+  return false;
+}
+
+static std::ostream &
+operator<<(std::ostream & out, const rmw_time_t & t)
+{
+  out << "{" << t.sec << "s " << t.nsec << "ns}";
+  return out;
+}
+
+static void
+_write_to_buffer(const std::string message, char * buffer, size_t buffer_size)
+{
+  // Only write if a buffer is provided
+  if (!buffer || buffer_size == 0u) {
+    return;
+  }
+  size_t write_size = message.size();
+  // Only write up to the buffers max size
+  if (write_size > buffer_size) {
+    write_size = buffer_size;
+  }
+  strncpy(buffer, message.c_str(), write_size);
+}
+
+rmw_ret_t
+qos_profile_check_compatible(
+  const rmw_qos_profile_t publisher_qos,
+  const rmw_qos_profile_t subscription_qos,
+  rmw_qos_compatibility_type_t * compatibility,
+  char * reason,
+  size_t reason_size)
+{
+  if (!compatibility) {
+    _write_to_buffer("ERROR: compatibility parameter is null", reason, reason_size);
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  std::ostringstream reason_ss;
```

Comment:
I see that the point of having a `reason` and `reason_size` arguments is to avoid allocations, but a `std::ostringstream` will actually allocate in the heap.

Reply:
you can instead use `rcutils_snprintf` like [here](https://github.com/ros2/rcutils/blob/1eae6891b2ef54563ab6d11d564ebdabc7939398/include/rcutils/error_handling.h#L235-L244).

Reply:
Good catch, I missed that!

Reply:
I've refactored to avoid using ostringstream in 9c43a6e